src/privatesend/privatesend-util.cpp

```diff
@@ -70,3 +76,244 @@ void CKeyHolderStorage::ReturnAll()
         LogPrintf("CKeyHolderStorage::%s -- %lld keys returned\n", __func__, tmp.size());
     }
 }
+
+CTransactionBuilderOutput::CTransactionBuilderOutput(CTransactionBuilder* pTxBuilderIn, CWallet* pwalletIn, CAmount nAmountIn) :
+    pTxBuilder(pTxBuilderIn),
+    key(pwalletIn),
+    nAmount(nAmountIn)
+{
+    assert(pTxBuilder);
+    CPubKey pubKey;
+    key.GetReservedKey(pubKey, false);
+    script = ::GetScriptForDestination(pubKey.GetID());
+}
+
+bool CTransactionBuilderOutput::UpdateAmount(const CAmount nNewAmount)
+{
+    LOCK(pTxBuilder->cs_outputs);
+
+    const CAmount nAmountDiff = nNewAmount - nAmount;
+
+    if (nNewAmount > 0 && nAmountDiff <= pTxBuilder->GetAmountLeft()) {
+        if (nAmountDiff != 0) {
+            nAmount = nNewAmount;
+        }
+        return true;
+    }
+    return false;
+}
+
+CTransactionBuilder::CTransactionBuilder(CWallet* pwalletIn, const CompactTallyItem& tallyItemIn) :
```

Comment:
All of this CTransactionBuilder stuff really needs unit tests

Reply:
I agree, it makes sense and i actually thought about doing it initially but i didn't because at the moment i probably wouldn't be able to provide them without a decent delay.. i first need to work myself into our unit test stuff and its best practices, never really looked at it yet and if i do it, i want to do it right and understand all of t and not just do something that its done. I have it on my list and i will come back to it once i get the time. Hope that works! But also for the meantime, feel free to provide some unit tests if you can manage to get them together easier.

Reply:
I want to hear Udjin's thoughts... This is a relatively complicated class we are adding, and I'm not sure I'd want this merged without at least basic unit tests